tx-pool/src/component/orphan.rs

```diff
@@ -98,46 +105,60 @@ impl OrphanPool {
             .cloned()
             .collect();
 
-        let mut evicted = expires.len();
+        let mut evicted_txs = vec![];
 
         for id in expires {
-            self.remove_orphan_tx(&id);
+            if let Some(entry) = self.remove_orphan_tx(&id) {
+                evicted_txs.push(entry.tx.hash());
+            }
         }
 
         while self.len() > DEFAULT_MAX_ORPHAN_TRANSACTIONS {
-            evicted += 1;
             // Evict a random orphan:
             let id = self.entries.keys().next().cloned().expect("bound checked");
-            self.remove_orphan_tx(&id);
+            if let Some(entry) = self.remove_orphan_tx(&id) {
+                evicted_txs.push(entry.tx.hash());
+            }
         }
 
-        if evicted > 0 {
-            trace!("OrphanTxPool full, evicted {} tx", evicted);
+        if !evicted_txs.is_empty() {
+            trace!("OrphanTxPool full, evicted {} tx", evicted_txs.len());
             self.shrink_to_fit();
         }
-        evicted
+        evicted_txs
     }
 
-    pub fn add_orphan_tx(&mut self, tx: TransactionView, peer: PeerIndex, declared_cycle: Cycle) {
+    pub fn add_orphan_tx(
+        &mut self,
+        tx: TransactionView,
+        peer: PeerIndex,
+        declared_cycle: Cycle,
+    ) -> Vec<Byte32> {
         if self.entries.contains_key(&tx.proposal_short_id()) {
-            return;
+            return vec![];
         }
-        debug!("add_orphan_tx {}", tx.hash());
 
+        debug!("add_orphan_tx {}", tx.hash());
         self.entries.insert(
             tx.proposal_short_id(),
             Entry::new(tx.clone(), peer, declared_cycle),
         );
 
         for out_point in tx.input_pts_iter() {
-            self.by_out_point.insert(out_point, tx.proposal_short_id());
+            self.by_out_point
+                .entry(out_point)
+                .or_insert_with(HashSet::default)
+                .insert(tx.proposal_short_id());
         }
-        self.limit_size();
+
+        self.limit_size()
     }
 
-    pub fn find_by_previous(&self, tx: &TransactionView) -> Option<ProposalShortId> {
+    pub fn find_by_previous(&self, tx: &TransactionView) -> Vec<ProposalShortId> {
```

Comment:
I think we may return `Vec<&ProposalShortId>` since it's only used to find orphan tx by id internally, it will remove the `clone` operation in this fn.